python/python/bystro/tada_mixture/mixture_tada_em.py

```diff
@@ -80,44 +83,114 @@ def fit(self, X, progress_bar=True, Lamb_init=None, pi_init=None):
+            for j in range(self.K):
+                log_pmf = st.poisson.logpmf(X, Lambda_[j])
+                log_pmf_sum = np.sum(log_pmf, axis=1)
+                z_probs[:, j] = log_pmf_sum + np.log(pi_[j])
+            expectation_z = np.argmax(z_probs, axis=1)
+
+            for j in range(self.K):
+                pi_[j] = np.mean(expectation_z == j)
+                Lambda_[j] = np.mean(X[expectation_z == j], axis=0)
+
+            eps = 0.001
+            pi_ = pi_ * (1 - self.K * eps) + np.ones(self.K) * eps
+
+            self._save_losses(
+                i,
+            )
+
+        self._store_instance_variables([Lambda_, pi_])
+        return self
+
+    def _store_instance_variables(self, trainable_variables):
+        """
+        Saves the learned variables
+
+        Parameters
+        ----------
+        trainable_variables : list
+            List of variables to save
+        """
+        self.pi_ = trainable_variables[1]
+        self.Lambda_ = trainable_variables[0]
+
+    def _initialize_variables(self, X, Lamb_init, pi_init):
+        """
+        This initializes the factor loadings (Lambda), and the component
+        probabilities (pi). If Lamb_init  or pi_init are given these are
+        the values obviously. Otherwise, the initialization of the weights
+        is random and the Loadings are a Gaussian mixture model.
+
+        Parameters
+        ----------
+
+        Returns
+        -------
+        """
         # Initialize variables
         if Lamb_init is None:
-            model = GaussianMixture(K)
+            model = GaussianMixture(self.K)
             model.fit(X)
             Lambda_ = model.means_
         else:
             Lambda_ = Lamb_init
 
         rng = np.random.default_rng(2021)
-        pi_ = rng.dirichlet(np.ones(K)) if pi_init is None else pi_init
+        pi_ = rng.dirichlet(np.ones(self.K)) if pi_init is None else pi_init
 
-        myrange = trange if progress_bar else range
+        return Lambda_, pi_
 
-        for i in myrange(td["n_iterations"]):
-            # E step compute most likely categories
-            z_probs = np.zeros((N, K))
-            for j in range(K):
-                log_pmf = st.poisson.logpmf(X, Lambda_[j])
-                log_pmf_sum = np.sum(log_pmf, axis=1)
-                z_probs[:, j] = log_pmf_sum + np.log(pi_[j])
-            Ez = np.argmax(z_probs, axis=1)
+    def _initialize_save_losses(self):
+        """
+        This method initializes the arrays to track relevant variables
+        during training
 
-            # M step, maximize parameters
-            for j in range(K):
-                pi_[j] = np.mean(Ez == j)
-                Lambda_[j] = np.mean(X[Ez == j], axis=0)
+        TODO
 
-            eps = 0.001
-            pi_ = pi_ * (1 - K * eps) + np.ones(K) * eps
+        Parameters
+        ----------
+        """
 
-        self.Lambda = Lambda_
-        self.pi = pi_
-        return self
+    def _save_losses(self,i):
```

Comment:
could you help me understand what's happening here?  It looks like this method isn't implemented, and we're not using the argument `i`, which ruff and / or mypy should be warning about